Translate List and Keys (#55)
reactjs/de.reactjs.org: content/docs/lists-and-keys.md

---
id: lists-and-keys
title: Listen und Keys
permalink: docs/lists-and-keys.html
prev: conditional-rendering.html
next: forms.html
---

Lass uns zuerst schauen, wie wir Listen in JavaScript transformieren können.

> In dem untenstehenden Code verwenden wir die [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) Funktion um die Werte eines `Nummern`-Arrays zu verdoppeln. Wir weisen das von `map()` neu zurückgegebene Array der Variable `double` zu und loggen es:

Reviewer: ```suggestion
In dem untenstehenden Code verwenden wir die [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)-Funktion um die Werte eines `Nummern`-Arrays zu verdoppeln. Wir weisen das von `map()` neu zurückgegebene Array der Variable `double` zu und loggen es:
```

Reviewer: Verwendest du konsequent `das` oder `der` Array?
[Duden](https://www.duden.de/rechtschreibung/Array) sagt beides möglich, ich nutze `der`.
Sollte letztendlich eben konsequent immer `das` oder `der` sein.

Author: Verwende immer `das`, da es für mich am natürlichsten klingt. Beispielsweise Sätze wie: `Das mit Nummern gefüllte Array`, hören sich irgendwie mit einem `der` ein wenig komisch an.

```javascript{2}
const numbers = [1, 2, 3, 4, 5];
const doubled = numbers.map((number) => number * 2);
console.log(doubled);
```

Dieser Code schreibt `[2, 4, 6, 8, 10]` in die Konsole.

> In React, ist das transformieren von Arrays in [Element](/docs/rendering-elements.html)-Listen nahezu identisch.

Reviewer: Das Komma als Betonung ist im Englischen gebräuchlich,
im Deutschen scheint es falsch zu sein: [Quelle](http://canoo.net/blog/2014/09/22/trotz-der-langen-adverbialbestimmung-am-satzanfang-steht-vor-steht-kein-komma)

> In React, ist das transformieren von Arrays in [Element](/docs/rendering-elements.html)-Listen nahezu identisch.

Author: Bin mir da auch immer ein wenig unsicher. In der ganzen Dokumentation befinden sich so viele überlange Sätze, die durch viele Kommata getrennt sind. Danke dir für die ganzen Quellen ;)

> ### Rendern meherer Komponenten {#rendering-multiple-components}

Reviewer: ```suggestion
### Rendern mehrerer Komponenten {#rendering-multiple-components}
```

Du kannst Sammlungen von Elementen erstellen und diese mit geschweiften Klammern `{}` [in JSX einbinden](/docs/introducing-jsx.html#embedding-expressions-in-jsx).

> Nachfolgend wird ein `Nummern`-Array mit Hilfe der JavaScript [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) Funktion. Wir geben für jeden Eintrag im Array ein `<li>` Element zurück. Schließlich weisen wir das Array der Variable `listItems` zu:

Reviewer: ```suggestion
Nachfolgend wird ein `Nummern`-Array mit Hilfe der JavaScript [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)-Funktion durchlaufen. Wir geben für jeden Eintrag im Array ein `<li>` Element zurück. Schließlich weisen wir das Array der Variable `listItems` zu:
```

);
```

> Wir fügen das gesamte `listItems` Array in ein `<ul>` Element, und [rendern es in das DOM](/docs/rendering-elements.html#rendering-an-element-into-the-dom):

Reviewer: ```suggestion
Wir fügen das gesamte `listItems`-Array in ein `<ul>`-Element und [rendern es in das DOM](/docs/rendering-elements.html#rendering-an-element-into-the-dom):
```

Reviewer: https://de.wikipedia.org/wiki/Durchkopplung

);
```

Wenn du keine IDs für die gerenderten Elemente hast, kannst du als letzte Maßnahme auch den Listenindex benutzen:

```js{2,3}
const todoItems = todos.map((todo, index) =>
  // Mache dies nur, wenn dir keine IDs zur Verfügung stehen
  <li key={index}>
    {todo.text}
  </li>
);
```

> Wir empfehlen dir nicht die Indizes für die Keys zu verwenden, da sich die Reihenfolge der Listeneinträge verändern kann. Dies kann sich negativ auf die Performance auswirken und zu Problemen mit dem Komponenten-State führen. Im Artikel von Robin Pokorny's kannst du eine [ausführlichere Erklärung über die negativen Auswirkungen beim Verwenden des Index als Key](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318) finden. Wenn du Listenelemente keinen expliziten Key zuweist, verwendet React standardmäßig die Indizes als Key.

Reviewer: ```suggestion
Wir empfehlen dir nicht die Indizes für die Keys zu verwenden, da sich die Reihenfolge der Listeneinträge verändern kann. Dies kann sich negativ auf die Performance auswirken und zu Problemen mit dem Komponenten-State führen. Im Artikel von Robin Pokorny kannst du eine [ausführlichere Erklärung über die negativen Auswirkungen beim Verwenden des Index als Key](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318) finden. Wenn du Listenelemente keinen expliziten Key zuweist, verwendet React standardmäßig die Indizes als Key.
```

);
```

> In dem obigen Beispiel, kann die `Post`-Komponente `props.id` lesen, aber nicht `props.key`.

Reviewer: ```suggestion
In dem obigen Beispiel kann die `Post`-Komponente `props.id` lesen, aber nicht `props.key`.
```

}
```

[**Auf CodePen ausprobieren**](https://codepen.io/gaearon/pen/BLvYrB?editors=0010)

> Manchmal führt dies zu klarerem Code, aber diese Stil kann ebenso missbruacht werden. Wie in JavaScript liegt es an dir zu entscheiden ob es sich lohnt eine Variable zur besseren Lesbarkeit zu benutzen. Behalte im Hinterkopf, dass wenn `map()` zu verschachtelt ist, es wahrscheinlich eine guter Zeitpunkt ist diesen Code in eine [Komponente auszulagern](/docs/components-and-props.html#extracting-components).

Reviewer: ```suggestion
Manchmal führt dies zu klarerem Code, aber dieser Stil kann ebenso missbraucht werden. Wie in JavaScript liegt es an dir zu entscheiden, ob es sich lohnt, eine Variable zur besseren Lesbarkeit zu benutzen. Behalte im Hinterkopf, dass wenn `map()` zu verschachtelt ist, es wahrscheinlich eine guter Zeitpunkt ist, diesen Code in eine [Komponente auszulagern](/docs/components-and-props.html#extracting-components).
```

Reviewer: Paar Kommata, damit der Satz einfacher zu lesen ist.
